### PR TITLE
Support correlation

### PR DESCRIPTION
**Pull request by mrproliu**
Support correlation custom data to tracing context.

### CODE REVIEW

**Comment by wu-sheng** [comment]
This PR breaks the existing propagation APIs, but it is reasonable. @kezhenxu94 @arugal @hanahmily please review the codes detail.

**Comment by wu-sheng** [comment]
We need to make the API stable before the 1.0 release.

**Comment by wu-sheng** [comment]
@mrproliu  I think this PR would break all plugins in https://github.com/SkyAPM/go2sky-plugins as most of them are going context propagation. You need to update those after this merges.

**Comment by mrproliu** [comment]
> @mrproliu I think this PR would break all plugins in https://github.com/SkyAPM/go2sky-plugins as most of them are going context propagation. You need to update those after this merges.

Of course. After this merge, I will update the plugins.

File: README.md [hunk 1]
```diff
@@ -67,6 +67,22 @@ A sub span created as the children of root span links to its parent with `Contex
...
+```go
+value := go2sky.GetCorrelation(ctx, key)
+```
+
+## Put correlation
+
+Put custom data to tracing context.
+
+```go
+go2sky.PutCorrelation(ctx, key, value)
```
Comment (hanahmily) [inline README.md:83]: I prefer to add a bool result to show whether putting is successful. Users could leverage it to throw a panic, log an error message or ignore it at all.

Comment (mrproliu) [reply README.md]: Thanks, has added.

File: correlation.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+func PutCorrelation(ctx context.Context, key, value string) bool {
+	activeSpan := ctx.Value(ctxKeyInstance)
+	if activeSpan == nil {
+		return false
+	}
+
+	span, ok := activeSpan.(segmentSpan)
+	if ok {
+		span.context().CorrelationContext[key] = value
```
Comment (wu-sheng) [inline correlation.go:30]: Take a look at Java's `Config.Correlation.*`. This propagation is not unlimited for keeping the whole system safe.

Comment (mrproliu) [reply correlation.go:30]: Should I add a `opt` to build the tracer? such as `WithCorrelation(keyCount, valueSize int)` ?

File: README.md [hunk 3]
```diff
@@ -109,16 +125,16 @@ upstream service.
 
 ```go
 //Extract context from HTTP request header `sw8`
-span, ctx, err := tracer.CreateEntrySpan(r.Context(), "/api/login", func() (string, error) {
-		return r.Header.Get("sw8"), nil
+span, ctx, err := tracer.CreateEntrySpan(r.Context(), "/api/login", func(key string) (string, error) {
```
Comment (arugal) [inline README.md:128]: All the plugin implementations need to be modified, if merged, right :)

Comment (arugal) [reply README.md:128]: Please ignore

File: correlation.go (hunk 2 above)
Comment (wu-sheng) [reply correlation.go:30]: Yes, for go sdk, configs are all from there, until we support CDS later.

Comment (mrproliu) [reply correlation.go]: Added. please take a look.

**Review by wu-sheng — APPROVED** [review]

**Comment by wu-sheng** [comment]
@mrproliu Please update the plugins.

**Pull request merged**